snowpark_style_guide.md

```diff
@@ -0,0 +1,341 @@
+being descriptive or too short lacking enough information. Keeping those long column names while manipulating `DataFrame`
+could somehow reduce maintainability and readability such that aliasing is wanted.
+
+There are two ways to do aliasing in Snowpark : either `DataFrame.with_column_renamed` or `Column.alias`
+suffices the purpose. `with_column_renamed` separates column selection and renaming.
+In contrast, `Column.alias` brings the benefit
+of locality as well as makes code more compact which further improves readability and conciseness.
+
+```python
+# with_column_renamed
+order = order.select(
+    'restaurant_id',
+    'order_id',
+    'order_type',
+    'order_customer_id',
+    'price',
+    'time'
+)
+order = order.with_column_renamed('order_customer_id', 'customer_id')
+order = order.with_column_renamed('price', 'order_price')
+order = order.with_column_renamed('time', 'order_time')
+
+# alias
+order = order.select(
+    'restaurant_id',
+    'order_id',
+    'order_type',
+    col('order_customer_id').alias('customer_id'),
+    col('cost').alias('order_cost'),
+    col('time').alias('order_time')
+)
+```
+
+## Performant SQL Generation
+
+Operations upon `DataFrame` will ultimately get turned into SQL statements and sent to Snowflake for execution.
+Non-optimized SQL statements would take longer time to process and execute and sometimes the service even times out.
+Snowpark has implemented a SQL simplifier which is turned on by default, and it is build upon certain rules.
+To maximize the effectiveness of the simplifier, we propose the following guidelines based on our practices:
+
+### Calling `DataFrame.with_columns` is preferred than `DataFrame.with_column`
+
+`with_column` will add another layer of `select` statement within the internal sql execution plan and keeps calling the
+function will ultimately generate which is too long to be executed. Instead of calling `DataFrame.with_column` for
+multiple times, please try using `DataFrame.with_columns` to batch your operations.
+
+```python
+# unperformant
+df = df.with_column("mean", (df["a"] + df["b"]) / 2)
+df = df.with_column("sum", df["a"] +  df["b"]).sort(df["a"]).show()
+
+# performant
+df = df.with_columns(
+   ["mean", "sum"],
+   [(df["a"] + df["b"]) / 2, df["a"] + df["b"]]
+).sort(df["a"]).show()
+```
+
+# Expressions
+
+Chaining of expressions are widely used as they represent atomic logic. Here we are give
+some best practice based on our experiences for writing the expressions.
+
+## Multiple-Lines
+
+In general, it should be avoided to lay out all the expression into a single line which makes code
+difficult to read with expressions left behind getting easier to be omitted.
+
+Wrapping the multi-lines expressions within parentheses provides a graceful way to laying out the code.
+
+```python
+# all in one line
+df = df.select('col_a', 'col_b').filter('col_a' == 'value').filter('cob_b' == True).join(another_dataframe, 'col_c').drop('col_d')
```

Comment:
```suggestion
df = df.select('col_a', 'col_b').filter(df['col_a'] == 'value').filter(df['cob_b'] == True).join(another_dataframe, 'col_c').drop('col_d')
```

Reply:
Should we call join with parameter names?

Reply:
good call out, updated